privacyidea/lib/eventhandler/requestmangler.py

```diff
@@ -0,0 +1,163 @@
+                       {"parameter":
+                            {"type": "str",
+                             "required": True,
+                             "description": _("The parameter that should be deleted.")}
+                        },
+                   ACTION_TYPE.ADD:
+                       {"parameter":
+                            {"type": "str",
+                             "required": True,
+                             "description": _("The parameter that should be added or modified.")
+                             },
+                        "value":
+                            {"type": "str",
+                             "required": True,
+                             "description": _("The new value of the parameter. Can contain tags like {0}, {1} for  "
+                                              "the matched sub strings.")
+                            },
+                        "match_parameter":
+                            {"type": "str",
+                             "description": _("The parameter, that should match some values.")
+                            },
+                        "match_pattern":
+                            {"type": "str",
+                             "description": _("The value of the match_parameter. It can contain a regular "
+                                              "expression and '()' to transfer values to the new parameter.")
+                            }
+                        }
+                   }
+        return actions
+
+    def do(self, action, options=None):
+        """
+        This method executes the defined action in the given event.
+
+        :param action:
+        :param options: Contains the flask parameters g, request, response
+            and the handler_def configuration
+        :type options: dict
+        :return:
+        """
+        ret = True
+        request = options.get("request")
+        handler_def = options.get("handler_def")
+        handler_options = handler_def.get("options", {})
+        parameter = handler_options.get("parameter")
+
+        if parameter:
+            if action.lower() == ACTION_TYPE.DELETE:
+                if parameter in request.all_data:
+                    del(request.all_data[parameter])
+            elif action.lower() == ACTION_TYPE.ADD:
+                value = handler_options.get("value")
+                match_parameter = handler_options.get("match_parameter")
+                match_pattern = handler_options.get("match_pattern")
+                if value:
+                    # We only take action, if we have a value
+                    if not match_parameter:
+                        # simple setting a parameter
+                        request.all_data[parameter] = value
+                    elif match_pattern:
+                        # setting a parameter depending on another value
+                        m = re.match(match_pattern, request.all_data.get(match_parameter))
```

Comment:
By default, match only searches for a match at the beginning of the match parameter value, i.e. it does not require the whole string to match. Should we maybe require ``match_pattern`` to match the whole string by passing ``"^" + match_pattern + "$"`` to ``re.match``? This seems safer to me.

Reply:
ok, i think I wanted to use ``re.search``, right?

Reply:
Just replacing ``re.match`` with ``re.search`` isn't right either because this searches for a match *anywhere* in the input. I only recently learned that ``re.match`` only requires a match at the beginning at the string -- any trailing characters *after* the match are just ignored.
So we either need to use ``re.match("^" + match_pattern + "$")`` or ``re.search("^" + match_pattern + "$")``, I think both would work.

Reply:
Hm, this is a question of definition. It is not defined, yet, that we always want to match the **complete** value of the match_parameter.

We could very well match a substring in the middle like: ``\w(.*)@(.*)\w`` to match something like an email address anywhere in the value.

The admin can decide to put the beinnging and ending "^", "$" in the match pattern himself.

My original intention was to use ``re.search``. Why shouldn't we?

Reply:
I think matching the whole string is a safer default: Assume we take the use ``re.match`` behavior, i.e. only look for a match at the beginning of the string. An admin wants to add a field "foo=bar" if the ``realm`` parameter is ``realm1``. The admin defines:
<pre>
parameter: foo
value: bar
match_parameter: realm
match_value: realm1
</pre>
This works for a while. Now the setup grows, and there is a new realm ``realm12``. Now, the rule above suddenly matches ``realm12`` as well, which is probably not intended by the admin.

If we require the pattern to match the whole string instead, the rule wouldn't match ``realm12``. In order to get the old ``re.match`` behavior, the admin could still define a pattern ``realm1.*``.

Reply:
Thank you. You convinced me!

Reply:
done in 5bd08dd5

Reply:
thanks!